legend-engine-ide-lsp-server/src/main/java/org/finos/legend/engine/ide/lsp/classpath/ClasspathUsingMavenFactory.java

```diff
@@ -209,28 +212,31 @@ public CompletableFuture<ClassLoader> create(Iterable<String> folders)
         });
     }
 
-    private void updatePlatformVersions(File maven, Path pom, File settingsXmlFile, List<SDLCPlatform> platforms)
+    private void updatePlatformVersions(Path pom, List<SDLCPlatform> platforms)
     {
         if (!platforms.isEmpty())
         {
+            this.server.logInfoToClient("Updating platform version on pom: " + pom);
             try
             {
-                String pomContent = Files.readString(pom);
+                Model model;
+                try (InputStream is = Files.newInputStream(pom))
+                {
+                    model = new MavenXpp3Reader().read(is);
+                }
 
                 for (SDLCPlatform platform : platforms)
                 {
                     String versionProperty = "platform." + platform.getName() + ".version";
-                    if (pomContent.contains(versionProperty))
-                    {
-                        Properties properties = new Properties();
-                        properties.setProperty("property", versionProperty);
-                        properties.setProperty("newVersion", platform.getPlatformVersion());
-                        properties.setProperty("generateBackupPoms", Boolean.FALSE.toString());
-                        this.invokeMaven(maven, pom.toFile(), settingsXmlFile, properties, "versions:set-property");
-                    }
+                    model.addProperty(versionProperty, platform.getPlatformVersion());
+                }
+
+                try (OutputStream os = Files.newOutputStream(pom))
+                {
+                    new MavenXpp3Writer().write(os, model);
```

Comment:
The maven Model uses the Properties class for maven properties, which does not maintain order. This means that the order of the properties can be changed between reading and writing, even if there's no change to property values. Which means you can end up with spurious diffs.

On top of that, I don't believe it's stable between multiple read/write cycles. So you can end up with constantly shifting property orderings, even if nothing is actually changing.

In legend-sdlc, we dealt with this by replacing the Properties objects with ones that sort the properties by key. But that's a slightly different case, since we don't expect manual edits to files. In this case, what you'd ideally like is to preserve the order that's already present.

Also, I would expect reading the model and writing it back like this is going to lose things like the original formatting and comments.

Finally, the formatting of MavenXpp3Writer isn't great. I personally am not a fan of it.

So I think I would recommend not using this method for updating the properties. I would even recommend something like regex-based replacement over this (even though that comes with plenty of problems of its own).

Reply:
This make sense.  In the current iteration I just committed, I moved to use regex.  Let me know what you think with this implementation.